pkg/diagnostics/http_monitoring.go

```diff
@@ -307,3 +346,94 @@ func (h *httpMetrics) HTTPMiddleware(next http.Handler) http.Handler {
 		h.ServerRequestCompleted(r.Context(), method, path, status, reqContentSize, respSize, elapsed)
 	})
 }
+
+func (h *httpMetrics) initPathNormalization(config *config.PathNormalization) pathNormalizationConfig {
+	if config == nil {
+		return pathNormalizationConfig{
+			enabled: false,
+		}
+	}
+
+	pn := pathNormalizationConfig{
+		enabled: config.Enabled,
+	}
+
+	if !pn.enabled {
+		return pn
+	}
+
+	pn.virtualIngressMux = http.NewServeMux()
+	pn.virtualIngressMux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !h.legacy {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = diagUtils.UnmatchedPath
+		}
+	}))
+	for _, pattern := range config.IngressPaths {
+		pn.virtualIngressMux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = pattern
+		}))
+	}
+
+	pn.virtualEgressMux = http.NewServeMux()
+	pn.virtualEgressMux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !h.legacy {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = diagUtils.UnmatchedPath
+		}
+	}))
+	for _, pattern := range config.EgressPaths {
+		pn.virtualEgressMux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*diagUtils.PathNormalizationRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathNormalizationRW")
+				return
+			}
+			rw.NormalizedPath = pattern
+		}))
+	}
+	return pn
+}
+
+func (h *httpMetrics) normalizePath(path string, direction int) (string, bool) {
+	if !h.pathNormalization.enabled {
+		return "", false
+	}
+
+	if path == "" {
+		return "", false
+	}
+
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+
+	req := &http.Request{
```

Comment:
you might want to benchmark with `sync.pool` to create this as well, to compare with/without it, as sync.pool might make sense in this use case